// viser-ng/lib/module.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Global = exports.registerShape = exports.registerAnimation = exports.ViserModule = void 0;

var _common = require("@angular/common");

var _core = require("@angular/core");

var viser = _interopRequireWildcard(require("viser"));

var _Chart = require("./Chart");

var _index = require("./components/index");

var _LiteChart = require("./LiteChart");

var _Plugin = require("./plugins/Plugin");

var _Slider = require("./plugins/Slider");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

var __decorate = void 0 && (void 0).__decorate || function (decorators, target, key, desc) {
  var c = arguments.length,
      r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc,
      d;
  if ((typeof Reflect === "undefined" ? "undefined" : _typeof(Reflect)) === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);else for (var i = decorators.length - 1; i >= 0; i--) {
    if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
  }
  return c > 3 && r && Object.defineProperty(target, key, r), r;
};

var ViserModule = function () {
  function ViserModule() {}

  ViserModule = __decorate([(0, _core.NgModule)({
    imports: [_common.CommonModule],
    declarations: [_Chart.Chart, _LiteChart.LiteChart, _index.Axis, _index.Brush, _index.Coord, _index.Facet, _index.Guide, _index.Legend, _index.Tooltip, _index.View, _index.FacetView, _index.Series, _index.Pie, _index.Sector, _index.Line, _index.SmoothLine, _index.DashLine, _index.Area, _index.StackArea, _index.SmoothArea, _index.Bar, _index.StackBar, _index.DodgeBar, _index.Point, _index.Funnel, _index.Pyramid, _index.Schema, _index.Box, _index.Candle, _index.Polygon, _index.Contour, _index.Heatmap, _index.Edge, _index.Sankey, _index.ErrorBar, _index.JitterPoint, _index.StackInterval, _index.DodgeInterval, _index.Interval, _index.Path, _index.Venn, _Plugin.PluginComponent, _Slider.Slider],
    exports: [_Chart.Chart, _LiteChart.LiteChart, _index.Axis, _index.Brush, _index.Coord, _index.Facet, _index.Guide, _index.Legend, _index.Tooltip, _index.View, _index.FacetView, _index.Series, _index.Pie, _index.Sector, _index.Line, _index.SmoothLine, _index.DashLine, _index.Area, _index.StackArea, _index.SmoothArea, _index.Bar, _index.StackBar, _index.DodgeBar, _index.Point, _index.Funnel, _index.Pyramid, _index.Schema, _index.Box, _index.Candle, _index.Polygon, _index.Contour, _index.Heatmap, _index.Edge, _index.Sankey, _index.ErrorBar, _index.JitterPoint, _index.StackInterval, _index.DodgeInterval, _index.Interval, _index.Path, _index.Venn, _Plugin.PluginComponent, _Slider.Slider]
  })], ViserModule);
  return ViserModule;
}();

exports.ViserModule = ViserModule;
var registerAnimation = viser.registerAnimation;
exports.registerAnimation = registerAnimation;
var registerShape = viser.registerShape;
exports.registerShape = registerShape;
var Global = viser.Global;
exports.Global = Global;
